mmverify.py

```diff
@@ -1,369 +1,588 @@
+                    i = stmt.index('$=')
+                    proof = stmt[i + 1:]
+                    stmt = stmt[:i]
+                except ValueError as exc:
+                    raise MMError('$p must contain a proof after $=') from exc
+                dvs, f_hyps, e_hyps, conclusion = self.fs.make_assertion(stmt)
                 if not self.begin_label:
-                    vprint(1, 'verifying', label)
-                    self.verify(label, stat, proof)
-                self.labels[label] = ('$p', self.fs.make_assertion(stat))
+                    vprint(2, 'verifying:', label)
+                    self.verify(f_hyps, e_hyps, conclusion, proof)
+                self.labels[label] = ('$p', (dvs, f_hyps, e_hyps, conclusion))
                 label = None
-            elif tok == '$d': self.fs.add_d(toks.readstat())
-            elif tok == '${': self.read(toks)
-            elif tok[0] != '$': label = tok
-            else: print('tok:', tok)
+            elif tok == '$d':
+                self.fs.add_d(toks.readstmt())
+            elif tok == '${':
+                self.read(toks)
+            elif tok[0] != '$':
+                label = tok
+            else:
+                print('Unknown token:', tok)
             tok = toks.readc()
         self.fs.pop()
 
-    def apply_subst(self, stat, subst):
-        result = []
-        for tok in stat:
-            if tok in subst: result.extend(subst[tok])
-            else: result.append(tok)
-        vprint(20, 'apply_subst', (stat, subst), '=', result)
-        return result
-
-    def find_vars(self, stat):
-        vars = []
-        for x in stat:
-            if not x in vars and self.fs.lookup_v(x): vars.append(x)
-        return vars
-
-    def decompress_proof(self, stat, proof):
-        dm, mand_hyp_stmts, hyp_stmts, stat = self.fs.make_assertion(stat)
-
-        mand_hyps = [self.fs.lookup_f(v) for k, v in mand_hyp_stmts]
-        hyps = [self.fs.lookup_e(s) for s in hyp_stmts]
-
-        labels = mand_hyps + hyps
-        hyp_end = len(labels)
-        ep = proof.index(')')
-        labels += proof[1:ep]
-        compressed_proof = ''.join(proof[ep+1:])
-
-        vprint(5, 'labels:', labels)
+    def treat_step(self,
+                   step: FullStmt,
+                   stack: list[Stmt]) -> None:
+        """Carry out the given proof step (given the label to treat and the
+        current proof stack).  This modifies the given stack in place.
+        """
+        steptyp, stepdat = step
+        vprint(10, 'step', steptyp, stepdat)
+        if steptyp in ('$e', '$f'):
+            stack.append(stepdat)
+        elif steptyp in ('$a', '$p'):
+            dvs0, f_hyps0, e_hyps0, conclusion0 = stepdat
+            vprint(12, stepdat)
+            npop = len(f_hyps0) + len(e_hyps0)
+            sp = len(stack) - npop
+            if sp < 0:
+                raise MMError('stack underflow')
+            subst: dict[Var, Stmt] = {}
+            for k, v in f_hyps0:
```

Comment:
I saw `k` being called "typecode" elsewhere, but you might want to keep `k` to keep naming convention with the previous version.